clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -1256,7 +1256,9 @@ private CompletedFetch initializeCompletedFetch(CompletedFetch nextCompletedFetc
                         log.debug("Discarding stale fetch response for partition {} since the fetched offset {} " +
                                 "does not match the current offset {}", tp, fetchOffset, subscriptions.position(tp));
                     } else if (subscriptions.hasDefaultOffsetResetPolicy()) {
-                        log.info("Fetch offset {} is out of range for partition {}, resetting offset", fetchOffset, tp);
+                        log.info("Fetch offset {} is out of range for partition {}. We only have log segments in " +
+                                "the range from {} to {}. Resetting offset", fetchOffset, tp, partition.logStartOffset,
+                                partition.lastStableOffset);
```

Comment:
The last stable offset is not the log end offset. We need to use the isolation level to decide whether to use the high watermark or the last stable offset as the upper bound. If the isolation level is READ_UNCOMMITTED, then we use high watermark; if READ_COMMITTED, then we use last stable offset.

Reply:
Refactored